### PR TITLE
[*] Performances : Try to improve cart performances (PSCSX-5943)

### PR DESCRIPTION
**Pull request by jocel1**

### CODE REVIEW

**Comment by studiokiwik** [comment]
no need for autoincrement, you can just use shuffled array's index from php; finally, it's all about the making benchmark what is faster, array_shuffle or ORDER BY RAND() for the typical and extreme cases (what is typical number of product_reductions?)

**Comment by jocel1** [comment]
If you shuffle the array without using ORDER BY, you don't have any guarantee the results will be returned in random order after running the query (it depends on the JOIN order chosen by MySQL).
In my case the array has about 1000 elements.

**Comment by studiokiwik** [comment]
I mean "ORDER BY product_reductions.sort_key" , where sort_key is index of array_shuffle($product_reductions)   (CREATE TEMPORARY TABLE product_reductions sort_key, id_product, id_product_attribute ...)
If I understand correctly, product_reductions table's scope is limited to script execution?

**Comment by jocel1** [comment]
Yes, the temp table exists only in the current db connection scope. Having sort_key or a autoincrement PK is almost the same here. If you want to take advantage of a quick ORDER BY, you need a BTREE index on the sort_key, otherwise MySQL will do a filesort, which will be equivalent to the ORDER BY RAND().

**Comment by kpodemski** [comment]
With these changes shopping cart with 75 unique products loads in 1.4s instead of 5s

:+1: !

**Comment by gRoussac** [comment]
:clap:

File: classes/Cart.php [hunk 1]
```diff
@@ -1819,9 +1814,9 @@ public function getPackageList($flush = false)
 			$id_warehouse = 0;
 			foreach ($warehouse_count_by_address[$product['id_address_delivery']] as $id_war => $val)
 			{
-				if (in_array((int)$id_war, $product['warehouse_list']))
+				if (array_key_exists((int)$id_war, $product['warehouse_list']))
 				{
-					$product['carrier_list'] = array_merge($product['carrier_list'], Carrier::getAvailableCarrierList(new Product($product['id_product']), $id_war, $product['id_address_delivery'], null, $this));
+					$product['carrier_list'] = array_replace($product['carrier_list'], Carrier::getAvailableCarrierList(new Product($product['id_product']), $id_war, $product['id_address_delivery'], null, $this));
```
Comment (jeromenadaud) [inline classes/Cart.php]: array_replace is php 5.3, we still assume compatibility with php 5.2, soon we will remove it, but for this moment...

Comment (guestisp) [inline classes/Cart.php]: Compatibility for a PHP version that was EOL five years ago ?

Comment (tchauviere) [inline classes/Cart.php]: @guestisp,

We are not the only one to do so out there:
https://wordpress.org/about/requirements/

I think you would be surprised to see how many people/hosting providers still use this PHP version.

Anyway we are upgrading to PHP 5.4 as of 1.6.1.0 release.

Don't forget to have a look at build.prestashop.com. This is were we write about PrestaShop future :)

**Pull request merged**